Refactor error handling for implicit actions

qnkhuat opened the pull request:
Closes https://github.com/metabase/metabase/issues/32694

We're working on making actions error more user-friendly ([thread](https://metaboat.slack.com/archives/C057T1QTB3L/p1690462817579789?thread_ts=1690451722.933519&cid=C057T1QTB3L))

In this PR I refactored the error handling flow with 2 main goals:
- Making defining SQL error message parser easier and unifying the shape of error data
- Reduce the amount of custom try-catch in the call stack

Additional to that:
- we add tests around to cover all error cases
- add a parser for h2 which didn't exist before
- updated the error message to be more friendly([thread](https://metaboat.slack.com/archives/C057T1QTB3L/p1692023106337609?thread_ts=1690451722.933519&cid=C057T1QTB3L))


```
- Violate not null
  - Summarize : Name, <field-X>, and Email must have values.
  - Per column : You must provide a value.
- Violate unique constraints
  - Summarize: URL and Short code already exist.
  - Per column: This URL value already exists.
- Violate incorrect type
  - Summarize: Some of your values aren’t of the correct type for the database.
  - Per column: This value needs to be <Text | a Number | a Date>
- Violate FK constraint
  - Delete
   - Summarize: Other tables rely on this record so it cannot be deleted.
   - Per column: N/A I think?
  - Create
   - Summarize: Unable to create a new record
   - Per column: This Product ID does not exist
  - Update
   - Summarize: Unable to update the record
   - Per column: This Product ID does not exist
```

qnkhuat committed:
Change shape of error returned

qnkhuat committed:
WIP

qnkhuat committed:
WIP

qnkhuat committed:
getting rid another layer of error throwing

qnkhuat committed:
it works on UI

qnkhuat committed:
parse the fk violate case correctly

qnkhuat committed:
avoid duplicating error msg

qnkhuat committed:
No more messages

qnkhuat committed:
test fixes

qnkhuat committed:
put back the tests

qnkhuat committed:
even better error msg

qnkhuat committed:
remove the old method

qnkhuat committed:
further simplify error handling logic

qnkhuat committed:
even more simplified and tests

qnkhuat committed:
remove debug code

qnkhuat committed:
even more simplified

qnkhuat committed:
simplify code to compilembql query

qnkhuat committed:
even more simplified

qnkhuat commented on src/metabase/actions/execution.clj:
```diff
@@ -159,11 +156,8 @@
-      (catch Exception e
-        (handle-action-execution-error e)))))
+    (binding [qp.perms/*card-id* (:model_id action)]
+      (actions/perform-action! implicit-action arg-map))))
```
I think we need a nice job catching and rethrowing errors inside `perform-action!`, there is no need to do another try catch here.

qnkhuat commented on src/metabase/driver/mysql/actions.clj:
```diff
@@ -52,67 +55,55 @@
+                                          "(conn=183) Incorrect integer value: 'STRING' for column `table`.`id` at row 1")
   nil)
 
-(defmethod sql-jdbc.actions/parse-sql-error :mysql
```
removing `sql-jdbc.actions/parse-sql-error` was my original motivation, I think having a list of functions for each driver like this is hard to know which drivers support which parsing error function.

qnkhuat commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -23,41 +24,68 @@
   [driver database e]
-  (let [default-method? (= (get-method parse-sql-error driver)
-                           (get-method parse-sql-error :default))]
-    (if default-method?
```
this logic was also the motivation for the refactor, it's really hard to know what will be parsed upstream and what's being processed.

qnkhuat committed:
remove unecessary when

metamben reviewed (commented):
I like the direction this is going.

tsmacdonald commented on src/metabase/driver/mysql/actions.clj:
```diff
@@ -52,67 +55,53 @@
+          table      (remove-backticks table)
+          columns    (constraint->column-names database table constraint)]
+      {:type    error-type
+       :message (tru "Value for column(s) {0} is duplicated" (str/join ", " columns))
```
+1, we can do it. Look at https://github.com/metabase/metabase/blob/34d895a30dac2832d38a48ef339da7e7ad5cb4ac/test/metabase/util/i18n_test.clj#L112 for inspiration

qnkhuat committed:
Update src/metabase/driver/sql_jdbc/actions.clj

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

qnkhuat committed:
Merge branch 'master' into ngoc-actions-error-handling-cleanup

qnkhuat commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -23,41 +24,68 @@
+          [actions.error/violate-unique-constraint
+           actions.error/violate-foreign-key-constraint
+           actions.error/violate-not-null-constraint
+           actions.error/incorrect-value-type])
```
> What if a driver has some other error type, shouldn't it be able to return it?

I've changed this static list to a method to get all error types given a driver. so any custom driver error parser should be able to return.

> What if a driver can easily tell what error type the message is about (e.g., the message starts with an error code)

Notice the try catch in `do-with-auto-parse-sql-error` catch the SQLException class, so if there are error that driver can recognize, just throw a normal exception, it'll not get parsed.

metamben replied:
If that's the suggested way, we should probably document it. I expect most driver authors would want to profit from the existing framework as much as possible and might not think about doing everything from "scratch".

It's much easier to understand and maintain a set of functions with clear interfaces than throwing the exceptions of the right shape and info. If we go and document the exception "interface" then it's not a problem though.

qnkhuat replied:
https://github.com/metabase/metabase/blob/bcb605ec5bee47b3a9d7e40eb9e0159e18712d84/src/metabase/driver/sql_jdbc/actions.clj#L34

is the docstring here state that well?

qnkhuat commented on src/metabase/driver/postgres/actions.clj:
```diff
@@ -2,22 +2,26 @@
+             (re-find #"ERROR:\s+(\w+) value in column \"([^\"]+)\" of relation \"([^\"]+)\" violates not-null constraint"  error-message)]
+    {:type    error-type
+     :message (tru "Value for column {0} must be not null" column)
+     :errors  {column (tru "The value must be not null")}}))
```
Could you elaborate? I don't understand what default implementation you want here.

> I expect this map to be the same across almost all DB types.

yes, that's basically the ide aof `maybe-parse-sql-error`, we expect it either returns `nil`, or a map with these 3 keys.

qnkhuat replied:
I see, I think this comment cover that https://github.com/metabase/metabase/pull/32935#discussion_r1293142525

TLDR is that I initially did something similar to that, but error messages from each driver does not contain the same level of granularity so it's hard to generalize these.

qnkhuat commented on src/metabase/actions/error.clj:
```diff
@@ -0,0 +1,17 @@
+
+(def incorrect-value-type
+  "Error type for SQL incorrect value type."
+  ::incorrect-value-type)
```
My motivation for defining var for these things are:
- we can provide docs
- easy to find all error types
- centralized and maybe in the future we can do sth like error hierarchy as we have in `metabase.query-processor.error-type`

I could remove this ns if you find it's not justified. normal keywords also work for me.

calherries replied:
It might be useful to add another layer of indirection and centralize the creation of the error maps currently returned by each implementation of `maybe-parse-sql-error`. For each error-type, we need to return the same format of data under the `:message` and `:errors` keys. We should only need to define the `:message` and error messages under `:errors` in one place for each error-type, not once for each combination of driver and error-type. Then the only thing you would need to define in each driver's maybe-parse-sql-error implementation is how to extract the required components from the DBs error message.

qnkhuat replied:
@calherries 
Initially, I went with that approach and I really hope we could.

but the problem is the error message from each driver does not provide the same level of granularity.

For example for incorrect type error.
Mysql error message: `Incorrect integer value: 'STRING' for column "table.id" at row 1`
Postgres error message: `invalid input syntax for type integer: \"S\"`

Notice the error message from Postgres does not contain which column it has type error. 
So if we want the message and errors to be the same for all drivers, it enforces the error message to be generic and requires as few components as possible.

That's why I changed to "error" per driver approach like this, it's extra effort, but it gives more flexibility.

qnkhuat committed:
dynamic error types

calherries commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -23,41 +24,72 @@
-                   more-info)
-            e)))
+  (try
+   (some #(maybe-parse-sql-error driver % database (ex-message e)) (driver->available-error-parser driver))
```
This could be simplified. `(methods maybe-parse-sql-error)` returns a dispatch table of driver and error-type to parser function. You're effectively looking up this table twice just to get the parser functions. Once to get the error-types for a driver, and then using those error-types, you get the methods. But you could just do it in one step, where you get the methods for a given driver and apply them directly.

e.g.

```clojure
(defn- parse-sql-error
  [driver database e]
  (let [parsers-for-driver (keep (fn [[[method-driver _error-type] method]]
                                   (when (= method-driver driver)
                                     method))
                                 (dissoc (methods maybe-parse-sql-error) :default))]
    (try
      (some #(% driver database (ex-message e)) parsers-for-driver)
      ;; Catch errors in parse-sql-error and log them so more errors in the future don't break the entire action.
      ;; We'll still get the original unparsed error message.
      (catch Throwable new-e
        (log/error new-e (trs "Error parsing SQL error message {0}: {1}" (pr-str (ex-message e)) (ex-message new-e)))
        nil))))
```

qnkhuat commented on src/metabase/driver/postgres/actions.clj:
```diff
@@ -27,38 +31,55 @@
+     :message (tru "Invalid value \"{0}\", expect type: {1}" value expected-type)
+     :errors  {}}))
+
+(comment
+ (sql-jdbc.actions/maybe-parse-sql-error
+  :postgres actions.error/violate-foreign-key-constraint {:id 47}
+  "ERROR: update or delete on table \"group\" violates foreign key constraint \"user_group-id_group_-159406530\" on table \"user\"\n  Detail: Key (id)=(1) is still referenced from table \"user\".")
+
+ (sql-jdbc.actions/maybe-parse-sql-error
+  :postgres actions.error/violate-unique-constraint {:id 47}
+  "Batch entry 0 UPDATE \"public\".\"group\" SET \"ranking\" = CAST(2 AS INTEGER) WHERE \"public\".\"group\".\"id\" = 1 was aborted: ERROR: duplicate key value violates unique constraint \"group_ranking_key\"\n  Detail: Key (ranking)=(2) already exists.  Call getNextException to see other errors in the batch.")
+
+ (sql-jdbc.actions/maybe-parse-sql-error
+  :postgres actions.error/violate-not-null-constraint nil
+  "ERROR: null value in column \"ranking\" of relation \"group\" violates not-null constraint\n  Detail: Failing row contains (57, admin, null).")
+
+ (sql-jdbc.actions/maybe-parse-sql-error
+  :postgres actions.error/incorrect-value-type nil
+  "Batch entry 0 UPDATE \"public\".\"group\" SET \"ranking\" = CAST('S' AS INTEGER) WHERE \"public\".\"group\".\"id\" = 1 was aborted: ERROR: invalid input syntax for type integer: \"S\"  Call getNextException to see other errors in the batch."))
```
I'll add more tests later, but I don't think we should have error strings as test cases, I want to go further and construct the actually query to throw such errors so that if message error changes between driver version, we can catch that and write more robust regex to cover those cases.

qnkhuat committed:
more efficient look up

qnkhuat committed:
make it more REPL friendly by using Cal's implementation

qnkhuat committed:
test fixes

qnkhuat commented:
Could anyone give another look at the overall shape? I'll polish the error message once we agree on the design.

calherries reviewed (approved):
Nice

qnkhuat committed:
improve docs

qnkhuat committed:
update error message template

qnkhuat committed:
adding more tests

qnkhuat committed:
tests for postgres and mysql should cover most cases

qnkhuat committed:
H2 error parser and tests

qnkhuat committed:
remove tray selects

qnkhuat commented on src/metabase/driver/mysql/actions.clj:
```diff
@@ -52,67 +55,53 @@
+       :errors  (reduce (fn [acc col]
+                          (assoc acc col (tru "This column has unique constraint and this value is existed")))
+                        {}
+                        columns)})))
```
we always have a summarized message for that, I think the intention of these per-column errors are to aid users in fixing the wrong value.

qnkhuat committed:
more fined tune fk error messages

qnkhuat committed:
test fixes

qnkhuat committed:
test fixes

qnkhuat commented:
@calherries I've updated the error messages and tests could you please do another review? You can find the spec in the PR's description

calherries commented on src/metabase/driver/mysql/actions.clj:
```diff
@@ -52,66 +40,82 @@
+             (re-find #"Duplicate entry '.+' for key '(.+)'" error-message)]
+    (let [column (last (str/split fk #"\."))]
+      {:type    error-type
+       :message (tru "{0} already exist." (str/capitalize column))
```
There's only one column, so this needs to be
```suggestion
       :message (tru "{0} already exists." (str/capitalize column))
```

qnkhuat replied:
good catch. Thanks

calherries commented on src/metabase/util/string.clj:
```diff
@@ -0,0 +1,28 @@
+   [metabase.util.i18n :refer [deferred-tru]]))
+
+
+(defn build-sentence
+  "Join parts of a sentence together to build a compound one.
+
+  Options:
+  - stop? (default true): whether to add a period at the end of the sentence
+
+  Examples:
+
+    (build-sentence [\"foo\" \"bar\" \"baz\"]) => \"foo, bar and baz.\"
+
+    (build-sentence [\"foo\" \"bar\" \"baz\"] :stop? false) => \"foo, bar and baz\"
+  "
+  ([parts]
+   (build-sentence parts :stop? true))
+  ([parts & {:keys [stop?]
+             :or   {stop? true}
+             :as _options}]
+   (when (seq parts)
+     (cond
+       (= (count parts) 1) (str (first parts) (when stop? \.))
+       (= (count parts) 2) (str (first parts) " " (deferred-tru "and")  " " (second parts) (when stop? \.))
+       :else               (str (first parts) ", " (build-sentence (rest parts)))))))
```
This doesn't work for all languages, especially right-to-left ones. And I'm pretty sure not all languages will use commas and "and" in that way to create lists. I'm not sure what the solution is, but this isn't it. Maybe someone has attempted to solve the problem before and has some clues, especially on the FE

qnkhuat replied:
https://shopify.engineering/internationalization-i18n-best-practices-front-end-developers

I read this the other day and I don't think they have a solution for dynamically building sentences.

Also,
1. Solving the generic problem of pluralization shouldn't be the goal of this PR
2. We only support 29 languages, 2 of which are right-to-left, so I think our pluralization is far from perfect. not that it's an excuse, but it supports the 1st point.
3. As this is exacted to be its own function, I think once we decided to solve this problem, we can solve it once for all.
4. Started a discussion here https://metaboat.slack.com/archives/C010L1Z4F9S/p1692241389613209

calherries replied:
Good points. Should we add a small FIXME at least explaining that this doesn't work for a lot of languages?

qnkhuat replied:
I added that as a note in the function https://github.com/metabase/metabase/blob/ngoc-actions-error-handling-cleanup/src/metabase/util/string.clj

calherries commented on src/metabase/driver/postgres/actions.clj:
```diff
@@ -27,38 +23,75 @@
+             (re-find #"duplicate key value violates unique constraint \"([^\"]+)\"" error-message)]
+    (let [columns (constraint->column-names database constraint)]
+      {:type    error-type
+       :message (tru "{0} already exist." (u/build-sentence (map str/capitalize columns) :stop? false))
```
This needs to be "exists" when there is one column. You can use `trun` instead of `tru` for that.

qnkhuat replied:
that's a bit tricky to use. [ecfde21](https://github.com/metabase/metabase/pull/32935/commits/ecfde213405235d7af83110b062b67654b1e87cf)

calherries commented on src/metabase/driver/h2/actions.clj:
```diff
@@ -21,3 +28,85 @@
+              :errors {column (tru "This {0} does not exist." (str/capitalize column))}}
+
+             :row/delete
+             {:message (tru "Other tables rely on this row so it cannot be updated/deleted.")
```
why updated?
```suggestion
             {:message (tru "Other tables rely on this row so it cannot be deleted.")
```

qnkhuat replied:
ah it should be removed. updated!

qnkhuat committed:
Properly pluralize

qnkhuat committed:
use deferred-trun and docs for build-sentence

qnkhuat committed:
make the regex more robust for h2

kamilmielnik commented:
> Closes #32935

It's this PR's own id. Closes #32694

qnkhuat committed:
add sql tests for each driver

qnkhuat committed:
slugified the columns

qnkhuat commented on src/metabase/driver/postgres/actions.clj:
```diff
@@ -27,38 +23,75 @@
+
+                 :row/update
+                 {:message (tru "Unable to update the record.")
+                  :errors  {column (tru "This {0} does not exist." (str/capitalize column))}})))
```
done!

kamilmielnik replied:
Works well, thanks!

qnkhuat committed:
remove unused ns

calherries commented on test/metabase/api/action_test.clj:
```diff
@@ -633,3 +639,99 @@
+                         (mt/user-http-request :rasta :post 400 (format "action/%d/execute" create-group)
+                                               {:parameters {"name" "admin" "ranking" nil}}))))
+                (testing "when updating"
+                  (is (= "Implicit parameters must be provided."
```
Is this because there's no way to update a value to be null?

calherries commented on test/metabase/api/action_test.clj:
```diff
@@ -633,3 +639,99 @@
+
+              (testing "violate unique constraint"
+                (testing "when creating"
+                  (is (= {:message "Ranking already exists."
```
We should probably have a test for the plural form, when there are multiple columns involved.

qnkhuat replied:
it's not easy to set up for that, we need to be able to define a not-null constraint for multiple columns, and it's currently not possible with our `defdataset`. 

it's good to cover that but I think it's too much effort.
I don't think we'll get that case a lot in the wild tho.

calherries replied:
I'm not sure it's that difficult to justify not covering it. We just need a table with composite PKs right? Here's an example.

https://github.com/metabase/metabase/blob/a47039c6b6d67aa811ae0d8e3f2c91195fe01e35/test/metabase/driver/mysql_test.clj#L448

You could copy+paste the code and change the json column to another type to make it easier.

qnkhuat commented on src/metabase/driver/h2/actions.clj:
```diff
@@ -29,6 +29,19 @@
+    (db-identifier->name \"PUBLIC.TABLE\" ) => \"TABLE\""
+  [s]
+  (-> s
+      (str/replace #"\"" "")
```
sorry but this replace the quote, not backslash. or am I missing something?

calherries replied:
Wait, is this the point of this line to unescape escaped quotes? If so explain with an example

calherries replied:
Ah, I'm mistaken. Never mind

kamilmielnik commented on src/metabase/actions/error.clj:
```diff
@@ -0,0 +1,17 @@
+(ns metabase.actions.error)
```
_leaving comment in an unrelated place_

I tried to execute an action with parameters that would violate 2 different FK constraints, but I'm only getting error for one of them:

![image](https://github.com/metabase/metabase/assets/6830683/4020069b-ebef-4c78-8f0a-250eaecd3f8f)

----

If I try to violate constraints separately, I correctly receive the detailed errors:

![image](https://github.com/metabase/metabase/assets/6830683/0c41ec4e-8fe1-4662-9eca-ee0594e63237)

qnkhuat commented on test/metabase/api/action_test.clj:
```diff
@@ -633,3 +639,99 @@
+     ["rasta"     2]
+     ["lucky"     1]]]])
+
+(deftest action-error-handling-test
```
no strong reason, we can refactor into 2 tests like that and can test the case `violate not-null constraint` when updating as well.

Will do tmr.

qnkhuat replied:
did this in [cc95f07](https://github.com/metabase/metabase/pull/32935/commits/cc95f072c399c1165df95478a655e597974483ee)

qnkhuat commented on test/metabase/api/action_test.clj:
```diff
@@ -633,3 +639,99 @@
+              (testing "violate not-null constraint"
+                (testing "when creating"
+                  (is (= {:message "Ranking must have values."
+                          :errors {:ranking "You must provide a value."}}
```
yea, `column-errors` might be better described that, but I have 2 counter points:
- errors are also used in the API to return `per param schema error`, so that's might be an existing pattern?
- our middleware have a special treatment for `:errors` though ([here](https://github.com/metabase/metabase/blob/ngoc-actions-error-handling-cleanup/src/metabase/server/middleware/exceptions.clj#L65)), so if we change to `column-errors` we'd need to treat it a bit different. Also this is an related [issue](https://github.com/metabase/metabase/issues/14683)

calherries replied:
Fair points, lets keep the `:errors` key. I wasn't aware that's how the middleware works.

qnkhuat committed:
Split tests to api and actions level

qnkhuat committed:
grammar

calherries commented on test/metabase/driver/h2_test.clj:
```diff
@@ -343,3 +346,53 @@
+            :h2 actions.error/violate-not-null-constraint nil :row/created
+            "NULL not allowed for column \"RANKING\"; SQL statement:\nINSERT INTO \"PUBLIC\".\"GROUP\" (\"NAME\") VALUES (CAST(? AS VARCHAR)) [23502-214])"))))
+
+  (testing "vioalte unique constraint"
```
bump

qnkhuat replied:
I fixed those in the new tests https://github.com/metabase/metabase/pull/32935/commits/cc95f072c399c1165df95478a655e597974483ee

calherries commented on test/metabase/driver/sql_jdbc/actions_test.clj:
```diff
@@ -51,3 +53,139 @@
+     (ex-data e))))
+
+(deftest action-error-handling-test
+  (mt/test-drivers (mt/normal-drivers-with-feature :actions)
```
I think this should be restricted to JDBC drivers for the future implementors of actions for non-jdbc drivers to skip these tests
```suggestion
  (mt/test-drivers (filter #(isa? driver/hierarchy % :sql-jdbc) (mt/normal-drivers-with-feature :actions))
```

kamilmielnik committed:
Show form-field-specific errors coming from execute action endpoints (#33258)

* Fix error passing in ActionParametersInputForm

* Allow any error to be returned from action

* Allow any error to be returned from dashboard action

* Add test for detailed errors

* Move test at the end of the suite

* Order fields to reflect the UI

* Use cy.button and introduce actionName

* Add test for detailed action errors in dashboard

* Revert import sort for better diff

* Revert import sort for a better diff

* Add tags to test

* Add assertion for multiple FK violations with explanatory comment

* Add test for detailed action errors in object details modal

* Decrease test flakiness

* Remove the 2 FK violations check as it is non-deterministic:
- locally, we'll get error for User id
- but in CI, we'll get the error for Product id instead

* Remove redundant navigation

* Make aliases consistent

* Shorten code

* Reuse createImplicitActions
- remove assertions unrelated to this test

* Remove unused intercept

kamilmielnik commented:
I've merged #33258 here which is the frontend part of form-field-level error handling when executing actions + e2e tests for it.

qnkhuat committed:
- Execute action error handling tests for sql dbs only
- Add a test case for unique constraint with multiple columns

calherries reviewed (approved):
LGTM, thanks for adding exhaustive tests

qnkhuat committed:
make sure we search constraint for mysql

qnkhuat committed:
Update mysql_test.clj

qnkhuat committed:
fix for mysql 5.7 and mariadb
